File: test-junit5/src/test/java/io/micronaut/test/junit5/TransactionalDefaultPropertyExplicitTest.java
package io.micronaut.test.junit5;

import io.micronaut.context.ApplicationContext;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import io.micronaut.test.support.TestPropertyProvider;
import io.micronaut.transaction.test.DefaultTestTransactionExecutionListener;
import jakarta.inject.Inject;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;

import java.util.Collections;
import java.util.Map;

@MicronautTest(transactional = true)
@io.micronaut.context.annotation.Property(name = "datasources.default.url", value = "jdbc:h2:mem:TransactionalDefaultPropertyExplicitTest;LOCK_TIMEOUT=10000;DB_CLOSE_ON_EXIT=FALSE")
@DbProperties
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class TransactionalDefaultPropertyExplicitTest implements TestPropertyProvider {

    @Inject
    ApplicationContext applicationContext;

    @Override
    public Map<String, String> getProperties() {
        return Collections.singletonMap("micronaut.test.transactional-default", "false");
    }

    @Test
    void transactionalEnabledWhenExplicit() {
        Assertions.assertTrue(applicationContext.containsBean(DefaultTestTransactionExecutionListener.class));
        Assertions.assertTrue(applicationContext.getEnvironment()
            .getProperty("micronaut.test.transactional", Boolean.class)
            .orElse(false));
    }
}

File: test-kotest5/src/test/kotlin/io/micronaut/test/kotest5/TransactionalDefaultPropertyExplicitTest.kt
package io.micronaut.test.kotest5

import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import io.micronaut.context.ApplicationContext
import io.micronaut.context.annotation.Property
import io.micronaut.test.extensions.kotest5.annotation.MicronautTest

@MicronautTest(transactional = true)
@Property(name = "micronaut.test.transactional-default", value = "false")
class TransactionalDefaultPropertyExplicitTest(
    private val applicationContext: ApplicationContext
) : StringSpec() {

    init {
        "explicit transactional annotation wins over default property" {
            applicationContext.environment
                .getProperty("micronaut.test.transactional", Boolean::class.java)
                .orElse(false) shouldBe true
        }
    }
}
